File: Soulo/Services/PatternDetectionService.swift
import Foundation

final class PatternDetectionService: Sendable {
    func detectPatterns(in entries: [JournalEntry]) async throws -> [DetectedPattern] {
        guard entries.count >= 3 else { return [] }

        return try await Task.detached(priority: .utility) {
            var patterns: [DetectedPattern] = []

            let sorted = entries.sorted { $0.timestamp < $1.timestamp }

            patterns += self.detectBrokenPromises(in: sorted)
            patterns += self.detectTopicAvoidance(in: sorted)
            patterns += self.detectSentimentDecline(in: sorted)
            patterns += self.detectGoalAbandonment(in: sorted)
            patterns += self.detectContradictions(in: sorted)
            patterns += self.detectCognitiveShift(in: sorted)
            patterns += self.detectRelationshipPatterns(in: sorted)
            patterns += self.detectDecisionPatterns(in: sorted)

            return patterns
                .filter { $0.severity >= UserDefaults.standard.integer(forKey: "insight_threshold").nonZero(default: 30) }
                .sorted { $0.severity > $1.severity }
        }.value
    }

    // MARK: - Broken Promise Detection

    private func detectBrokenPromises(in entries: [JournalEntry]) -> [DetectedPattern] {
        let commitmentPhrases = [
            "i will", "i'll", "i'm going to", "i need to", "i should",
            "i promise", "i swear", "starting", "going to start"
        ]
        let backtrackPhrases = ["i didn't", "i haven't", "i failed", "i couldn't", "still haven't"]

        var patterns: [DetectedPattern] = []
        var pendingCommitments: [(phrase: String, date: TimeInterval, entryId: String)] = []

        for entry in entries {
            let transcript = entry.transcript?.lowercased() ?? ""
            guard !transcript.isEmpty else { continue }

            for phrase in commitmentPhrases where transcript.contains(phrase) {
                if let range = transcript.range(of: phrase) {
                    let afterPhrase = transcript[range.upperBound...].trimmingCharacters(in: .whitespaces)
                    let commitment = String(afterPhrase.prefix(60))
                        .components(separatedBy: CharacterSet(charactersIn: ".!?\n")).first ?? ""

                    if commitment.count > 5 {
                        pendingCommitments.append((commitment, entry.timestamp, entry.id))
                    }
                }
            }

            for phrase in backtrackPhrases where transcript.contains(phrase) {
                pendingCommitments.removeAll { pending in
                    let daysSince = (entry.timestamp - pending.date) / 86400
                    let words = pending.phrase.split(separator: " ")
                    let matched = words.filter { transcript.contains($0.lowercased()) }.count
                    return daysSince > 1 && daysSince < 90 && matched >= words.count / 2
                }
            }
        }

        for commitment in pendingCommitments {
            let daysSince = (Date().timeIntervalSince1970 - commitment.date) / 86400
            guard daysSince > 3 else { continue }

            let severity = min(95, Int(daysSince) * 3 + 10)
            patterns.append(DetectedPattern(
                patternType: .brokenPromise,
                severity: severity,
                title: "Unkept Commitment",
                message: "You said you would '\(commitment.phrase)' \(Int(daysSince)) days ago. You haven't mentioned following through.",
                firstDetected: commitment.date,
                lastDetected: commitment.date,
                occurrenceCount: 1
            ))
        }

        return patterns
    }

    // MARK: - Topic Avoidance Detection

    private func detectTopicAvoidance(in entries: [JournalEntry]) -> [DetectedPattern] {
        guard entries.count >= 5 else { return [] }

        var topics: [String: [Double]] = [:] // topic -> list of vocal energies

        for entry in entries {
            guard let biomarkers = entry.biomarkers, let topicsList = entry.topics else { continue }

            for topic in topicsList {
                let energy = biomarkers.vocalEnergy
                topics[topic.topic, default: []].append(energy)
            }
        }

        let globalMeanEnergy = topics.values.flatMap { $0 }.reduce(0, +) / Double(max(1, topics.values.flatMap { $0 }.count))

        var patterns: [DetectedPattern] = []
        for (topic, energies) in topics where energies.count >= 2 {
            let topicMean = energies.reduce(0, +) / Double(energies.count)
            let energyDrop = globalMeanEnergy - topicMean

            if energyDrop > 0.2 { // Energy drops >20% on this topic
                let severity = min(90, Int(energyDrop * 150))
                patterns.append(DetectedPattern(
                    patternType: .topicAvoidance,
                    severity: severity,
                    title: "Topic Avoidance: \(topic)",
                    message: "Your vocal energy drops \(Int(energyDrop * 100))% when discussing \(topic). Mentioned \(energies.count) times with consistent avoidance.",
                    firstDetected: entries.first(where: { $0.topics?.contains(where: { $0.topic == topic }) == true })?.timestamp ?? Date().timeIntervalSince1970,
                    lastDetected: entries.last(where: { $0.topics?.contains(where: { $0.topic == topic }) == true })?.timestamp ?? Date().timeIntervalSince1970,
                    occurrenceCount: energies.count
                ))
            }
        }

        return patterns
    }

    // MARK: - Sentiment Decline Detection

    private func detectSentimentDecline(in entries: [JournalEntry]) -> [DetectedPattern] {
        guard entries.count >= 5 else { return [] }

        var topicSentiments: [String: [(date: TimeInterval, sentiment: Double)]] = [:]

        for entry in entries {
            guard let topics = entry.topics else { continue }
            for topic in topics {
                topicSentiments[topic.topic, default: []].append((entry.timestamp, topic.sentiment))
            }
        }

        var patterns: [DetectedPattern] = []
        for (topic, dataPoints) in topicSentiments where dataPoints.count >= 3 {
            let sortedData = dataPoints.sorted { $0.date < $1.date }
            let sentiments = sortedData.map { $0.sentiment }

            // Linear regression slope
            let n = Double(sentiments.count)
            let sumX = (0..<sentiments.count).reduce(0, +)
            let sumY = sentiments.reduce(0, +)
            let sumXY = sentiments.enumerated().map { Double($0.offset) * $0.element }.reduce(0, +)
            let sumX2 = (0..<sentiments.count).map { Double($0 * $0) }.reduce(0, +)

            let slope = (n * sumXY - Double(sumX) * sumY) / (n * sumX2 - Double(sumX) * Double(sumX))

            if slope < -0.05 { // Declining sentiment
                let severity = min(95, Int(abs(slope) * 200))
                let first = sortedData.first!.sentiment
                let last = sortedData.last!.sentiment
                let daysSpan = Int((sortedData.last!.date - sortedData.first!.date) / 86400)

                patterns.append(DetectedPattern(
                    patternType: .sentimentDecline,
                    severity: severity,
                    title: "Declining: \(topic)",
                    message: "Your sentiment about \(topic) has declined from \(String(format: "%.1f", first)) to \(String(format: "%.1f", last)) over \(daysSpan) days.",
                    firstDetected: sortedData.first!.date,
                    lastDetected: sortedData.last!.date,
                    occurrenceCount: sentiments.count
                ))
            }
        }

        return patterns
    }

    // MARK: - Goal Abandonment Detection

    private func detectGoalAbandonment(in entries: [JournalEntry]) -> [DetectedPattern] {
        let startPhrases = ["starting", "beginning", "new goal", "new habit", "this time", "motivated"]
        let declinePhrases = ["skipped", "missed", "hard", "difficult", "tired", "no time", "busy"]
        let abandonPhrases = ["gave up", "quit", "stopped", "failed", "couldn't do it", "not for me"]

        var patterns: [DetectedPattern] = []
        var activeGoals: [(id: UUID, goal: String, startDate: TimeInterval, entries: [String])] = []

        for entry in entries {
            let transcript = entry.transcript?.lowercased() ?? ""
            guard !transcript.isEmpty else { continue }

            // Detect goal start
            for phrase in startPhrases where transcript.contains(phrase) {
                let goal = extractGoal(from: transcript, after: phrase)
                if !goal.isEmpty {
                    activeGoals.append((UUID(), goal, entry.timestamp, [transcript]))
                }
            }

            // Track progress
            for idx in activeGoals.indices {
                let goalWords = activeGoals[idx].goal.split(separator: " ").map(String.init)
                let matchCount = goalWords.filter { transcript.contains($0.lowercased()) }.count
                if matchCount >= goalWords.count / 3 {
                    activeGoals[idx].entries.append(transcript)
                }
            }

            // Detect abandonment
            for phrase in abandonPhrases where transcript.contains(phrase) {
                for goal in activeGoals {
                    let goalWords = goal.goal.split(separator: " ").map(String.init)
                    let matchCount = goalWords.filter { transcript.contains($0.lowercased()) }.count
                    if matchCount >= goalWords.count / 3 {
                        let daysActive = (entry.timestamp - goal.startDate) / 86400
                        let severity = min(90, 30 + Int(daysActive) * 2)

                        patterns.append(DetectedPattern(
                            patternType: .goalAbandonment,
                            severity: severity,
                            title: "Abandoned Goal: \(goal.goal)",
                            message: "You started '\(goal.goal)' but mentioned giving up \(Int(daysActive)) days later. You mentioned it \(goal.entries.count) times with declining enthusiasm.",
                            firstDetected: goal.startDate,
                            lastDetected: entry.timestamp,
                            occurrenceCount: goal.entries.count
                        ))
                    }
                }
            }
        }

        // Detect cycles (multiple abandoned goals with similar patterns)
        if patterns.count >= 2 && Set(patterns.map { $0.title }).count >= 2 {
            let avgDaysToAbandon = patterns.map { Int(($0.lastDetected - $0.firstDetected) / 86400) }.reduce(0, +) / patterns.count
            if patterns.count >= 3 {
                let cyclePattern = DetectedPattern(
                    patternType: .goalAbandonment,
                    severity: max(patterns.map { $0.severity }.max() ?? 50, 70),
                    title: "Goal Abandonment Cycle",
                    message: "You've started and abandoned \(patterns.count) goals. Average time to abandon: \(avgDaysToAbandon) days. The pattern is nearly identical each time.",
                    firstDetected: patterns.first!.firstDetected,
                    lastDetected: patterns.last!.lastDetected,
                    occurrenceCount: patterns.count
                )
                return [cyclePattern]
            }
        }

        return patterns
    }

    // MARK: - Contradiction Detection

    private func detectContradictions(in entries: [JournalEntry]) -> [DetectedPattern] {
        guard entries.count >= 3 else { return [] }

        var topicSentiments: [String: [(date: TimeInterval, sentiment: Double, text: String)]] = [:]

        for entry in entries {
            guard let topics = entry.topics else { continue }
            let transcript = entry.transcript ?? ""
            for topic in topics {
                topicSentiments[topic.topic, default: []].append((entry.timestamp, topic.sentiment, transcript))
            }
        }

        var patterns: [DetectedPattern] = []
        for (topic, dataPoints) in topicSentiments where dataPoints.count >= 2 {
            let sorted = dataPoints.sorted { $0.date < $1.date }

            for i in 0..<sorted.count - 1 {
                for j in (i + 1)..<sorted.count {
                    let delta = abs(sorted[i].sentiment - sorted[j].sentiment)
                    if delta > 1.2 && abs(sorted[j].date - sorted[i].date) < 86400 * 30 {
                        let severity = min(85, Int(delta * 60))
                        patterns.append(DetectedPattern(
                            patternType: .contradiction,
                            severity: severity,
                            title: "Contradiction: \(topic)",
                            message: "You felt \(sentimentLabel(sorted[i].sentiment)) about \(topic), then \(sentimentLabel(sorted[j].sentiment)) within \(Int((sorted[j].date - sorted[i].date) / 86400)) days.",
                            firstDetected: sorted[i].date,
                            lastDetected: sorted[j].date,
                            occurrenceCount: 2
                        ))
                    }
                }
            }
        }

        return patterns
    }

    // MARK: - Cognitive Shift Detection

    private func detectCognitiveShift(in entries: [JournalEntry]) -> [DetectedPattern] {
        guard entries.count >= 10 else { return [] }

        let biomarkers: [(date: TimeInterval, speechRate: Double, hesitationRate: Double, pitchInstability: Double)] = entries.compactMap { entry in
            guard let b = entry.biomarkers, entry.biomarkersStatus == .done else { return nil }
            return (entry.timestamp, b.speechRate, b.hesitationRate, b.pitchInstability)
        }.sorted { $0.date < $1.date }

        guard biomarkers.count >= 10 else { return [] }

        let midpoint = biomarkers.count / 2
        let firstHalf = biomarkers[0..<midpoint]
        let secondHalf = biomarkers[midpoint..<biomarkers.count]

        func avg(_ keyPath: KeyPath<(date: TimeInterval, speechRate: Double, hesitationRate: Double, pitchInstability: Double), Double>) -> Double {
            let first = firstHalf.reduce(0) { $0 + $1[keyPath: keyPath] } / Double(firstHalf.count)
            let second = secondHalf.reduce(0) { $0 + $1[keyPath: keyPath] } / Double(secondHalf.count)
            return second - first
        }

        let speechRateDelta = avg(\.speechRate)
        let hesitationDelta = avg(\.hesitationRate)
        let pitchDelta = avg(\.pitchInstability)

        var patterns: [DetectedPattern] = []

        if abs(speechRateDelta) > 0.5 || abs(hesitationDelta) > 0.1 || abs(pitchDelta) > 20 {
            var changes: [String] = []
            if speechRateDelta > 0.5 { changes.append("speech rate increased \(String(format: "%.1f", speechRateDelta)) wps") }
            if speechRateDelta < -0.5 { changes.append("speech rate decreased \(String(format: "%.1f", abs(speechRateDelta))) wps") }
            if hesitationDelta > 0.1 { changes.append("hesitation increased \(String(format: "%.0f", hesitationDelta * 100))%") }
            if pitchDelta > 20 { changes.append("pitch instability increased") }

            let severity = min(90, Int(abs(speechRateDelta) * 30 + abs(hesitationDelta) * 100 + abs(pitchDelta) / 2))
            patterns.append(DetectedPattern(
                patternType: .cognitiveShift,
                severity: severity,
                title: "Speech Pattern Shift Detected",
                message: "Your speech patterns have changed: \(changes.joined(separator: ", ")). This could indicate increased stress, fatigue, or cognitive load over the past \(biomarkers.count / max(1, Int(biomarkers.last!.date - biomarkers.first!.date) / 86400)) days.",
                firstDetected: biomarkers.first!.date,
                lastDetected: biomarkers.last!.date,
                occurrenceCount: biomarkers.count
            ))
        }

        return patterns
    }

    // MARK: - Relationship Pattern Detection

    private func detectRelationshipPatterns(in entries: [JournalEntry]) -> [DetectedPattern] {
        let relationshipTerms = [
            "partner", "boyfriend", "girlfriend", "husband", "wife", "spouse",
            "mom", "dad", "mother", "father", "sister", "brother", "friend",
            "boss", "manager", "colleague", "team", "relationship", "marriage"
        ]

        var patterns: [DetectedPattern] = []
        var personMentions: [String: [(date: TimeInterval, sentiment: Double)]] = [:]

        for entry in entries {
            guard let topics = entry.topics else { continue }
            for topic in topics {
                let lower = topic.topic.lowercased()
                if relationshipTerms.contains(lower) || relationshipTerms.contains(where: { lower.contains($0) }) {
                    personMentions[topic.topic, default: []].append((entry.timestamp, topic.sentiment))
                }
            }
        }

        for (person, mentions) in personMentions where mentions.count >= 3 {
            let sorted = mentions.sorted { $0.date < $1.date }
            let avgSentiment = sorted.map { $0.sentiment }.reduce(0, +) / Double(sorted.count)

            if avgSentiment < -0.3 {
                let daysActive = Int((sorted.last!.date - sorted.first!.date) / 86400)
                let severity = min(85, Int(abs(avgSentiment) * 80))
                patterns.append(DetectedPattern(
                    patternType: .relationshipPattern,
                    severity: severity,
                    title: "Strained Relationship: \(person)",
                    message: "Your sentiment about \(person) is consistently negative (\(String(format: "%.1f", avgSentiment)) avg). Mentioned \(mentions.count) times over \(daysActive) days.",
                    firstDetected: sorted.first!.date,
                    lastDetected: sorted.last!.date,
                    occurrenceCount: mentions.count
                ))
            }

            // Volatility detection
            let volatility = sorted.enumerated().dropFirst().map { abs($0.element.sentiment - sorted[$0.offset - 1].sentiment) }.reduce(0, +) / Double(sorted.count - 1)
            if volatility > 0.8 {
                let severity = min(80, Int(volatility * 60))
                patterns.append(DetectedPattern(
                    patternType: .relationshipPattern,
                    severity: severity,
                    title: "Volatile Sentiment: \(person)",
                    message: "Your feelings about \(person) swing dramatically (volatility: \(String(format: "%.1f", volatility))). Mentioned \(mentions.count) times.",
                    firstDetected: sorted.first!.date,
                    lastDetected: sorted.last!.date,
                    occurrenceCount: mentions.count
                ))
            }
        }

        return patterns
    }

    // MARK: - Decision Outcome Detection

    private func detectDecisionPatterns(in entries: [JournalEntry]) -> [DetectedPattern] {
        guard entries.count >= 3 else { return [] }

        let sorted = entries.sorted { $0.timestamp < $1.timestamp }
        var decisions = DecisionOutcomeService.shared.scanDecisions(in: sorted)
        _ = DecisionOutcomeService.shared.detectFollowUps(in: sorted, decisions: &decisions)

        let regretPatterns = DecisionOutcomeService.shared.detectRegretPatterns(in: entries)
        var patterns = regretPatterns

        let regretted = decisions.filter { $0.status == .regretted }
        for decision in regretted {
            guard let followUpSentiment = decision.followUpSentiment else { continue }
            let severity = min(80, 30 + decision.daysSinceDecision * 2 + Int(abs(followUpSentiment) * 20))
            patterns.append(DetectedPattern(
                patternType: .decisionRegret,
                severity: severity,
                title: "Regretted Decision",
                message: "You regretted '\(decision.decisionText.prefix(60))' after \(decision.daysSinceDecision) days. Sentiment at outcome: \(String(format: "%.1f", followUpSentiment)).",
                dataJson: (try? String(data: JSONEncoder().encode(decision.id), encoding: .utf8)),
                firstDetected: decision.firstMentioned,
                lastDetected: decision.lastMentioned,
                occurrenceCount: decision.mentionCount
            ))
        }

        let pending = decisions.filter { $0.status == .pending && $0.daysSinceDecision > 14 }
        for decision in pending {
            let severity = min(50, 20 + decision.daysSinceDecision)
            patterns.append(DetectedPattern(
                patternType: .decisionRegret,
                severity: severity,
                title: "Unresolved Decision",
                message: "You decided '\(decision.decisionText.prefix(60))' \(decision.daysSinceDecision) days ago with no follow-up recorded.",
                dataJson: (try? String(data: JSONEncoder().encode(decision.id), encoding: .utf8)),
                firstDetected: decision.firstMentioned,
                lastDetected: decision.lastMentioned,
                occurrenceCount: decision.mentionCount
            ))
        }

        return patterns
    }

    // MARK: - Helpers

    private func extractGoal(from text: String, after phrase: String) -> String {
        guard let range = text.range(of: phrase) else { return "" }
        let afterPhrase = text[range.upperBound...].trimmingCharacters(in: .whitespaces)
        let goal = String(afterPhrase.prefix(80))
            .components(separatedBy: CharacterSet(charactersIn: ".!?\n")).first ?? ""
        return goal.trimmingCharacters(in: .whitespaces)
    }

    private func sentimentLabel(_ value: Double) -> String {
        if value > 0.5 { return "positively" }
        if value > 0.1 { return "somewhat positively" }
        if value < -0.5 { return "very negatively" }
        if value < -0.1 { return "negatively" }
        return "neutrally"
    }
}

File: android/app/src/main/java/com/soulo/app/services/PatternDetectionService.kt
package com.soulo.app.services

import com.soulo.app.models.*

object PatternDetectionService {
    private const val MIN_ENTRIES = 3
    private const val SENTIMENT_THRESHOLD = -0.3

    fun detectPatterns(
        entries: List<JournalEntry>,
        existingPatterns: List<DetectedPattern> = emptyList()
    ): List<DetectedPattern> {
        if (entries.size < MIN_ENTRIES) return emptyList()
        val sorted = entries.sortedBy { it.timestamp }
        val patterns = mutableListOf<DetectedPattern>()

        detectBrokenPromises(sorted, patterns)
        detectTopicAvoidance(sorted, patterns)
        detectSentimentDecline(sorted, patterns)
        detectGoalCycles(sorted, patterns)
        detectContradictions(sorted, patterns)
        detectCognitiveShift(sorted, patterns)
        detectRelationshipPatterns(sorted, patterns)
        detectDecisionRegret(sorted, patterns)

        return mergeWithExisting(patterns, existingPatterns)
    }

    private fun detectBrokenPromises(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val commitmentPhrases = listOf(
            "I will", "I'm going to", "I need to", "I should",
            "I promise", "I swear", "starting tomorrow", "this time"
        )
        val failurePhrases = listOf(
            "didn't", "couldn't", "failed", "gave up", "quit",
            "didn't follow through", "dropped the ball"
        )

        for (i in 1 until entries.size) {
            val prev = entries[i - 1].transcript?.lowercase() ?: continue
            val curr = entries[i].transcript?.lowercase() ?: continue
            val hasCommitment = commitmentPhrases.any { prev.contains(it) }
            val hasFailure = failurePhrases.any { curr.contains(it) }
            if (hasCommitment && hasFailure) {
                patterns.add(
                    DetectedPattern(
                        id = "bp_${entries[i].id}",
                        type = PatternType.brokenPromise,
                        title = "Broken Promise Detected",
                        message = "You committed to something on ${entries[i-1].formattedDate} but mentioned not following through.",
                        confidence = 0.7,
                        firstDetected = entries[i - 1].timestamp,
                        lastDetected = entries[i].timestamp,
                        occurrenceCount = 1
                    )
                )
            }
        }
    }

    private fun detectTopicAvoidance(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val topicFrequency = mutableMapOf<String, MutableList<Triple<Long, Double, Long>>>()
        for (entry in entries) {
            val topics = entry.topics ?: continue
            for (t in topics) {
                topicFrequency.getOrPut(t.topic) { mutableListOf() }
                    .add(Triple(entry.timestamp, t.sentiment, entry.durationMs))
            }
        }
        for ((topic, mentions) in topicFrequency) {
            if (mentions.size < 2) continue
            val sortedMentions = mentions.sortedBy { it.first }
            for (i in 1 until sortedMentions.size) {
                val gap = sortedMentions[i].first - sortedMentions[i - 1].first
                val prevDuration = sortedMentions[i - 1].third
                val currDuration = sortedMentions[i].third
                if (gap > 7 * 86400 && sortedMentions[i].second < 0 && currDuration < prevDuration * 0.7) {
                    patterns.add(
                        DetectedPattern(
                            id = "ta_${topic}_${sortedMentions[i].first}",
                            type = PatternType.topicAvoidance,
                            title = "Topic Avoidance: $topic",
                            message = "You mentioned '$topic' after a ${gap / 86400} day gap, but spoke about it briefly.",
                            confidence = 0.6,
                            relatedTopic = topic,
                            firstDetected = sortedMentions[i - 1].first,
                            lastDetected = sortedMentions[i].first
                        )
                    )
                }
            }
        }
    }

    private fun detectSentimentDecline(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val sorted = entries
            .filter { it.emotion != null && it.transcriptStatus == ProcessingStatus.completed }
            .sortedBy { it.timestamp }
        if (sorted.size < 3) return

        val recent = sorted.takeLast(3)
        val prevValence = sorted[sorted.size - 4].emotion!!.valence
        val recentAvg = recent.map { it.emotion!!.valence }.average()
        val decline = recentAvg - prevValence

        if (decline < SENTIMENT_THRESHOLD) {
            patterns.add(
                DetectedPattern(
                    id = "sd_${sorted.last().id}",
                    type = PatternType.sentimentDecline,
                    title = "Sentiment Decline",
                    message = "Your emotional valence dropped ${String.format("%.2f", kotlin.math.abs(decline))} points recently.",
                    confidence = kotlin.math.abs(decline).coerceAtMost(1.0),
                    firstDetected = sorted.first().timestamp,
                    lastDetected = sorted.last().timestamp
                )
            )
        }
    }

    private fun detectGoalCycles(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val goalPhrases = listOf(
            "I want to", "my goal", "I'm trying", "working on",
            "need to improve", "I should start", "I've been meaning to"
        )
        val abandonPhrases = listOf(
            "stopped", "gave up", "failed again", "can't keep up",
            "missed", "forgot", "didn't have time", "let it slip"
        )

        val goalMentions = mutableMapOf<String, MutableList<Pair<Long, String>>>()
        for (entry in entries) {
            val text = entry.transcript?.lowercase() ?: continue
            for (phrase in goalPhrases) {
                if (text.contains(phrase)) {
                    val goal = extractGoal(text, phrase)
                    goalMentions.getOrPut(goal) { mutableListOf() }
                        .add(entry.timestamp to phrase)
                }
            }
        }

        for ((goal, mentions) in goalMentions) {
            if (mentions.size < 3) continue
            val sortedM = mentions.sortedBy { it.first }
            var cycleCount = 0
            for (i in 1 until sortedM.size) {
                if (sortedM[i].second in abandonPhrases ||
                    sortedM[i - 1].second in abandonPhrases
                ) cycleCount++
            }
            if (cycleCount >= 2) {
                patterns.add(
                    DetectedPattern(
                        id = "gc_$goal",
                        type = PatternType.goalCycle,
                        title = "Goal Cycle: $goal",
                        message = "You've revisited '$goal' with ${mentions.size} mentions and $cycleCount gaps.",
                        confidence = 0.65,
                        firstDetected = sortedM.first().first,
                        lastDetected = sortedM.last().first,
                        occurrenceCount = cycleCount
                    )
                )
            }
        }
    }

    private fun extractGoal(text: String, phrase: String): String {
        val idx = text.indexOf(phrase) + phrase.length
        val end = text.indexOf('.', idx).takeIf { it > 0 } ?: text.indexOf(',', idx).takeIf { it > 0 } ?: (idx + 40).coerceAtMost(text.length)
        return text.substring(idx, end).trim()
    }

    private fun detectContradictions(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val contradictionPairs = listOf(
            "I love" to "I hate",
            "I'm happy" to "I'm miserable",
            "it's fine" to "it's terrible",
            "I'm good" to "I'm struggling",
            "no problem" to "big problem"
        )
        for (i in 1 until entries.size) {
            val prev = entries[i - 1].transcript?.lowercase() ?: continue
            val curr = entries[i].transcript?.lowercase() ?: continue
            for ((pos, neg) in contradictionPairs) {
                if (prev.contains(pos) && curr.contains(neg)) {
                    patterns.add(
                        DetectedPattern(
                            id = "ct_${entries[i].id}",
                            type = PatternType.contradiction,
                            title = "Contradictory Statements",
                            message = "You said '$pos' recently, but now mention '$neg'.",
                            confidence = 0.5,
                            firstDetected = entries[i - 1].timestamp,
                            lastDetected = entries[i].timestamp
                        )
                    )
                }
            }
        }
    }

    private fun detectCognitiveShift(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val sorted = entries
            .filter { it.biomarkers != null && it.emotion != null }
            .sortedBy { it.timestamp }
        if (sorted.size < 5) return

        val early = sorted.take(3)
        val late = sorted.takeLast(3)
        val earlyAvgSpeech = early.mapNotNull { it.biomarkers?.speechRate }.average()
        val lateAvgSpeech = late.mapNotNull { it.biomarkers?.speechRate }.average()
        val speechDelta = lateAvgSpeech - earlyAvgSpeech

        if (kotlin.math.abs(speechDelta) > 0.5) {
            val changes = mutableListOf<String>()
            if (speechDelta > 0.5) changes.add("speaking faster")
            if (speechDelta < -0.5) changes.add("speaking slower")
            if (late.map { it.biomarkers!!.pitchInstability }.average() > 0.3) changes.add("more pitch variation")

            if (changes.isNotEmpty()) {
                patterns.add(
                    DetectedPattern(
                        id = "cs_${sorted.last().id}",
                        type = PatternType.cognitiveShift,
                        title = "Cognitive Shift Detected",
                        message = "Your speech patterns have changed: ${changes.joinToString(", ")}.",
                        confidence = 0.55,
                        firstDetected = sorted.first().timestamp,
                        lastDetected = sorted.last().timestamp
                    )
                )
            }
        }
    }

    private fun detectRelationshipPatterns(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val people = mutableSetOf<String>()
        val relationshipTerms = listOf(
            "partner", "spouse", "boyfriend", "girlfriend", "husband", "wife",
            "mother", "father", "mom", "dad", "sister", "brother", "friend",
            "manager", "colleague", "therapist", "doctor"
        )

        for (entry in entries) {
            val text = entry.transcript?.lowercase() ?: continue
            for (term in relationshipTerms) {
                if (text.contains(term)) {
                    people.add(term)
                }
            }
        }

        for (person in people.take(5)) {
            val mentions = entries.filter { it.transcript?.lowercase()?.contains(person) == true }
            if (mentions.size < 2) continue
            val sentiments = mentions.mapNotNull { it.emotion?.valence }
            if (sentiments.size < 2) continue
            val trend = sentiments.zipWithNext().map { (a, b) -> b - a }
            val avgDelta = if (trend.isNotEmpty()) trend.average() else 0.0
            val volatility = if (sentiments.isNotEmpty()) {
                val mean = sentiments.average()
                sentiments.map { kotlin.math.abs(it - mean) }.average()
            } else 0.0

            if (volatility > 0.4) {
                patterns.add(
                    DetectedPattern(
                        id = "rp_$person",
                        type = PatternType.relationshipPattern,
                        title = "Relationship Pattern: $person",
                        message = "Your sentiment around '$person' shows high volatility (${String.format("%.2f", volatility)}).",
                        confidence = volatility.coerceAtMost(1.0),
                        firstDetected = mentions.first().timestamp,
                        lastDetected = mentions.last().timestamp,
                        occurrenceCount = mentions.size
                    )
                )
            }
        }
    }

    private fun detectDecisionRegret(entries: List<JournalEntry>, patterns: MutableList<DetectedPattern>) {
        val regretPhrases = listOf("I regret", "I shouldn't have", "I wish I hadn't", "big mistake", "wrong choice", "I made a mistake")
        for (entry in entries) {
            val text = entry.transcript?.lowercase() ?: continue
            val matched = regretPhrases.firstOrNull { text.contains(it) } ?: continue
            patterns.add(
                DetectedPattern(
                    id = "dr_${entry.id}",
                    type = PatternType.decisionRegret,
                    title = "Decision Regret",
                    message = "You expressed regret: \"$matched\"",
                    confidence = 0.8,
                    firstDetected = entry.timestamp,
                    lastDetected = entry.timestamp,
                    relatedEmotions = listOfNotNull(entry.emotion?.primaryEmotion)
                )
            )
        }
    }

    private fun mergeWithExisting(
        newPatterns: List<DetectedPattern>,
        existing: List<DetectedPattern>
    ): List<DetectedPattern> {
        val merged = existing.toMutableList()
        for (np in newPatterns) {
            val match = merged.find {
                it.type == np.type && it.relatedTopic == np.relatedTopic && it.title == np.title
            }
            if (match != null) {
                val idx = merged.indexOf(match)
                merged[idx] = match.copy(
                    occurrenceCount = match.occurrenceCount + 1,
                    lastDetected = np.lastDetected,
                    confidence = maxOf(match.confidence, np.confidence)
                )
            } else {
                merged.add(np)
            }
        }
        return merged
    }
}
